concept/type_/type_manager/validation/commit_time_validation.rs

```diff
@@ -4,18 +4,919 @@
+
+        for relates in relates_declared {
+            let role = relates.role();
+
+            Self::validate_role_is_unique_for_relation_type_hierarchy(
+                type_manager,
+                snapshot,
+                relation_type.clone(),
+                role.clone(),
+                validation_errors,
+            )?;
+            Self::validate_type_ordering(type_manager, snapshot, role.clone(), validation_errors)?;
+            Self::validate_type_annotations(type_manager, snapshot, role.clone(), validation_errors)?;
+        }
+
+        Ok(())
+    }
+
+    fn validate_object_type<T: ObjectTypeAPI<'static> + KindAPI<'static>>(
+        type_manager: &TypeManager,
+        snapshot: &impl ReadableSnapshot,
+        type_: T,
+        validation_errors: &mut Vec<SchemaValidationError>,
+    ) -> Result<(), ConceptReadError> {
+        let ill_owns = produced_errors!(
+            validation_errors,
+            Self::validate_declared_owns(type_manager, snapshot, type_.clone(), validation_errors)?
+        );
+        if !ill_owns {
+            Self::validate_abstractness_matches_with_capability::<Plays<'static>>(
+                type_manager,
+                snapshot,
+                type_.clone().into_owned_object_type(),
+                validation_errors,
+            )?;
+            Self::validate_capability_cardinality::<Owns<'static>>(
+                type_manager,
+                snapshot,
+                type_.clone(),
+                validation_errors,
+            )?;
+        }
+
+        let ill_plays = produced_errors!(
+            validation_errors,
+            Self::validate_declared_plays(type_manager, snapshot, type_.clone(), validation_errors)?
+        );
+        if !ill_plays {
+            Self::validate_abstractness_matches_with_capability::<Plays<'static>>(
+                type_manager,
+                snapshot,
+                type_.clone().into_owned_object_type(),
+                validation_errors,
+            )?;
+            Self::validate_capability_cardinality::<Plays<'static>>(
+                type_manager,
+                snapshot,
+                type_.clone(),
+                validation_errors,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn validate_declared_owns<T>(
+        type_manager: &TypeManager,
+        snapshot: &impl ReadableSnapshot,
+        type_: T,
+        validation_errors: &mut Vec<SchemaValidationError>,
+    ) -> Result<(), ConceptReadError>
     where
-        Snapshot: ReadableSnapshot,
+        T: ObjectTypeAPI<'static> + KindAPI<'static>,
     {
-        // TODO: See https://github.com/vaticle/typedb/pull/6981/ for an outline on validation
-        Ok(vec![])
+        let owns_declared: HashSet<Owns<'static>> = TypeReader::get_capabilities_declared(snapshot, type_.clone())?;
+
+        for owns in owns_declared {
+            Self::validate_overridden_owns(type_manager, snapshot, owns.clone(), validation_errors)?;
```

Comment:
The logic is different for relates and (owns/plays), so I decided to leave everything separated. Could combine `owns` and `plays`, but it's not a big deal to me.